TASK-47630 Introduce paragraphs for performances Technics Best practices (#32)
exoplatform/developer.exoplatform.org: content/best_practices.adoc

== Performance

> The reviewer must try to detect if the fix/feature could have significant bad impacts on the performances of the application. Performances issues can be expensive to find and fix, so it is important to raise any concern on this topic at this stage. Here are some examples of performance issues causes: too many database requests, slow database query, missing index in the database, too many HTTP requests, …

Reviewer: ```suggestion
The reviewer must try to detect if the fix/feature could have significant bad impacts on the performance of the application. Performances issues can be expensive to find and fix, so it is important to raise any concern on this topic at this stage. Here are some examples of performance issues causes: too many database requests, slow database queries, missing index in the database, too many HTTP requests, …
```

=== Front-End Performances

> ​In Front-End developing, it's important to consider best solutions that reduces Memory & CPU consumption in Browser. In fact, a Web page executes only one single thread for all Vue apps, so it's important to avoid adding a lot of stuff to do when mounting a Vue application (https://www.red-gate.com/simple-talk/development/dotnet-development/javascript-single-threaded/[Read more about JS lifecycle]). In fact, when adding a heavy computing statements in page loading phase, it will *pause* mounting all other Vue applications until the computing is finished, because of single-threaded execution architecture of Web pages. Consequently, having one Vue app that isn't sufficiently performant in a web page, especially for a TopBar application (that is present in all pages), will degrade the rendering performances of the whole page.
> 

Reviewer: ```suggestion
​In Front-End development, it's important to consider the best solutions that reduce Memory & CPU consumption in Browser. In fact, a Web page executes only one single thread for all Vue apps, so it's important to avoid adding a lot of stuff to do when mounting a Vue application (https://www.red-gate.com/simple-talk/development/dotnet-development/javascript-single-threaded/[Read more about JS lifecycle]). In fact, when adding heavy computing statements in the page loading phase, it will *pause* mounting all other Vue applications until the computing is finished, because of the single-threaded execution architecture of Web pages. Consequently, having one Vue app that isn't sufficiently performant in a web page, especially for a TopBar application (that is present on all pages), will degrade the rendering performances of the whole page.

```

> Knowing that, the Vue applications rendering has to be made in an asynchrnous way without impacting other applications present in the page. For example, when having information to retrieve from Server-End using REST API calls, a fetch call will be used. To ensure to have a good UX, a loading effect in the application itself has to be rendered first time the Vue application is mounted, until all needed information are retrieved.
> 
> But, using the fetch API to retrieve information asynchronously from Server-End has a big drawback. In fact, a Browser allows a limited concurrent HTTP requests for each domain (https://blog.stackpath.com/glossary-domain-sharding/[Domain Sharding]). In a web page, that holds a lot of Vue applications, each one will retrieve information from Server-End using different REST endpoints. This can consequently lead to a very bad UX with a considerable idle time for HTTP requests. Fortunately, the Web Standards has evolved to take into consideration this complexity of new Web Applications especially by introducing the `HTTP/2` protocol (https://www.limelightonline.co.nz/blog/what-http2-benefit-your-website/[Read more about benifits]). To benefit from it, and to allow to increase the parallel concurrent requests to send to Server-End, we will have to `preload` and `prefetch` mandatory information that renders the Web Page (https://medium.com/reloading/preload-prefetch-and-priorities-in-chrome-776165961bbf[Read more]).

Reviewer: ```suggestion
Knowing that the Vue applications rendering has to be made in an asynchronous way without impacting other applications present on the page. For example, when having information to retrieve from Server-End using REST API calls, a fetch call will be used. To ensure to have a good UX, a loading effect in the application itself has to be rendered the first time the Vue application is mounted, until all needed information is retrieved.

But, using the fetch API to retrieve information asynchronously from Server-End has a big drawback. In fact, a Browser allows limited concurrent HTTP requests for each domain (https://blog.stackpath.com/glossary-domain-sharding/[Domain Sharding]). In a web page, that holds a lot of Vue applications, each one will retrieve information from Server-End using different REST endpoints. This can consequently lead to a very bad UX with a considerable idle time for HTTP requests. Fortunately, the Web Standards has evolved to take into consideration this complexity of new Web Applications especially by introducing the `HTTP/2` protocol (https://www.limelightonline.co.nz/blog/what-http2-benefit-your-website/[Read more about benifits]). To benefit from it, and to allow to increase the parallel concurrent requests to send to Server-End, we will have to `preload` and `prefetch` mandatory information that renders the Web Page (https://medium.com/reloading/preload-prefetch-and-priorities-in-chrome-776165961bbf[Read more]).
```

> 
> The retrieved HTML DOM of eXo Platform contains references to the needed CSS files (computed from the list of Portlets being part of the displayed page). The fact that we have CSS links, in the retrieved DOM, will help browser to determine which files to retrieve using `HTTP/2` in an early stage of Web Page rendering phase. But the Javascript files uses a different mechanism based on https://requirejs.org/[RequireJS], that will retrieve JS files at the end of Web Page processing. This will lead to send multiple HTTP requests that will not be optimized using `HTTP/2`. Thus, the JS and CSS files are preloaded to ensure retrieving those files once at the same time, even when having tens of them. Consequently, all JS files has to use https://docs.exoplatform.org/en/6.1/JavaScript.html#amd-and-requirejs[AMD] mechanism to be imported inside a page and any *new module has to be imported as a dependency of a Portlet asset* to ensure not losing time in Web Page rendering phase. So, there are several *bad practices* to avoid such as, adding a `link` element inside a Vue application, or even worse when using `requirejs` in render phase on a module that is not a dependency of a displayed Portlet.

Reviewer: ```suggestion

The retrieved HTML DOM of eXo Platform contains references to the needed CSS files (computed from the list of Portlets being part of the displayed page). The fact that we have CSS links, in the retrieved DOM, will help the browser to determine which files to retrieve using `HTTP/2` in an early stage of the Web Page rendering phase. But the Javascript files use a different mechanism based on https://requirejs.org/[RequireJS], which will retrieve JS files at the end of Web Page processing. This will lead to sending multiple HTTP requests that will not be optimized using `HTTP/2`. Thus, the JS and CSS files are preloaded to ensure retrieving those files once at the same time, even when having tens of them. Consequently, all JS files have to use https://docs.exoplatform.org/en/6.1/JavaScript.html#amd-and-requirejs[AMD] mechanism to be imported inside a page and any *new module has to be imported as a dependency of a Portlet asset* to ensure not losing time in the Web Page rendering phase. So, there are several *bad practices* to avoid such as, adding a `link` element inside a Vue application, or even worse when using `requirejs` in the render phase on a module that is not a dependency of a displayed Portlet.
```

> 
> The `preload` and `prefetch` mechanisms can be used for other type of HTTP requests. In fact, each Vue application needs assets to be fully displayed. In general case, we will need:
> 
> * A CSS file HTTP request for application stylesheet: handled by Portal API to be preloaded, no specific modifications to make while developing a Vue application
> * A JS file HTTP request to execute Vue application rendering: handled by Portal API to be preloaded, no specific modifications to make while developing a Vue application
> * An I18N JSON file HTTP request: handled by Portal API to be preloaded, no specific modifications to make while developing a Vue application. This type of assets is handled by a different caching mechanism (https://developer.mozilla.org/fr/docs/Web/API/Window/sessionStorage[sessionStorage]) to improve the number of concurrent requests sent to the server (https://github.com/Meeds-io/commons/commit/cd66bda943e70cca43332de6940cacdf5cbde963[see details])

Reviewer: ```suggestion

The `preload` and `prefetch` mechanisms can be used for other types of HTTP requests. In fact, each Vue application needs assets to be fully displayed. In the general case, we will need:

* A CSS file HTTP request for application stylesheet: handled by Portal API to be preloaded, no specific modifications to make while developing a Vue application
* A JS file HTTP request to execute Vue application rendering: handled by Portal API to be preloaded, no specific modifications to make while developing a Vue application
* An I18N JSON file HTTP request: handled by Portal API to be preloaded, no specific modifications to make while developing a Vue application. This type of asset is handled by a different caching mechanism (https://developer.mozilla.org/fr/docs/Web/API/Window/sessionStorage[sessionStorage]) to improve the number of concurrent requests sent to the server (https://github.com/Meeds-io/commons/commit/cd66bda943e70cca43332de6940cacdf5cbde963[see details])
```

> * A set of HTTP requests to retrieve user settings and/or initial data to display
> 
> A Vue application can need one or several HTTP requests to retrieve data from Server-End. It's important to determine the priority of those requests and in which Web-Page rendering lifecycle, the request has to be sent. In fact, for the main application (The application that the user needs to retrieve its information, by example the Stream application in `/stream` page), the requests has to have a higher priority comparing to permanent page applications (TopBar applications). In order to increase the priority of a HTTP request, a `<link rel="preload"` or `<link rel="prefetch"` can be used to increase the priority. Even more, you can add a https://github.com/Meeds-io/social/blob/9549ffb38c861c080c1983e4fbf0c04dc8e33f58/webapp/portlet/src/main/webapp/WEB-INF/jsp/activityStream.jsp#L26[HTTP Header] that will allow to retrieve the HTTP request at the same time than JS & CSS files (highest priority just after Document DOM downloading). *BUT*, the preloaded resources shouldn't be slow queries, else the Web-Page rendering can be blocked for a while and consequently the Web-Page performances will degrade. As an example, the Stream page preload the list of Activity Identifiers without content to let Stream application load each activity apart. In order to improve activity loading, HTTP caching Headers has been used to efficiently cache activity content retrieving (Using eTag).

Reviewer: ```suggestion
* A set of HTTP requests to retrieve user settings and/or initial data to display

A Vue application can need one or several HTTP requests to retrieve data from Server-End. It's important to determine the priority of those requests and in which Web-Page rendering lifecycle, the request has to be sent. In fact, for the main application (The application that the user needs to retrieve its information, for example, the Stream application on `/stream` page), the requests have to have a higher priority comparing to permanent page applications (TopBar applications). In order to increase the priority of an HTTP request, a `<link rel="preload"` or `<link rel="prefetch"` can be used to increase the priority. Even more, you can add a https://github.com/Meeds-io/social/blob/9549ffb38c861c080c1983e4fbf0c04dc8e33f58/webapp/portlet/src/main/webapp/WEB-INF/jsp/activityStream.jsp#L26[HTTP Header] that will allow retrieving the HTTP request at the same time as JS & CSS files (highest priority just after Document DOM downloading). *BUT*, the preloaded resources shouldn't be slow queries, else the Web-Page rendering can be blocked for a while, and consequently, the Web-Page performances will degrade. As an example, the Stream page preloads the list of Activity Identifiers without content to let the Stream application load each activity apart. In order to improve activity loading, HTTP caching Headers have been used to efficiently cache activity content retrieving (Using eTag).
```

> Sometimes, when not having a lot of interactions with user on UI, (like for Wallet & Perkstore widgets), it's more relevant to compute the HTML to display in Server-End using a simple JSP file without having to mount a Vue application for it and to retrieve its information using additional REST calls. Using Vue applications is relevant only when :
> 

Reviewer: ```suggestion
Sometimes, when not having a lot of interactions with the user on UI, (like for Wallet & Perkstore widgets), it's more relevant to compute the HTML to display in Server-End using a simple JSP file without having to mount a Vue application for it and to retrieve its information using additional REST calls. Using Vue applications is relevant only when :

```

> * the computed information to display is slow: in this case, using a REST call to asynchronously retrieve information from Server-End will allow to detach heavy computing from critical path of Web-Page DOM retrieving
> * there are complex UI components and user interactions in application DOM
> 
> As a conclusion, to ensure having a good Front-End performances when developing a Vue application, we have multiple techniques that can be applied, but there is no exact coding pattern that we can follow. Each application has its specifities and performances requirement has to be considered in conception phase, else you can looase time to refactor code. In order to help developers to continuously measure developed/maintained applications performances, a tooling has been added to display application performances in Browser console when the server is started in https://docs.exoplatform.org/en/6.1/GetStarted.html#dev-mode[Dev Mode]. To enable this, use:

Reviewer: ```suggestion
* the computed information to display is slow: in this case, using a REST call to asynchronously retrieve information from Server-End will allow detaching heavy computing from the critical path of Web-Page DOM retrieving
* there are complex UI components and user interactions in application DOM

In conclusion, to ensure having good Front-End performances when developing a Vue application, we have multiple techniques that can be applied, but there is no exact coding pattern that we can follow. Each application has its specificities and performances requirement has to be considered in the conception phase, else you can lose time to refactor code. In order to help developers continuously measure developed/maintained applications performances, tooling has been added to display application performances in the Browser console when the server is started in https://docs.exoplatform.org/en/6.1/GetStarted.html#dev-mode[Dev Mode]. To enable this, use:
```

> 
> * `Vue.createApp` to create a new Vue application instead of `new Vue`
> * use statement `this.$root.$applicationLoaded()` when all Data is fetched and the UI has been displayed to the End-user.
> 
> === REST API Performances
> 
> By increasing the number of REST calls in a single Web-Page, this will allow to retrieve information in parallel. This was a major evolution of eXo Platform product architecture introduced in version 6 which will reduce the classic Server-End stateful DOM rendering (that was made using JSP, Servlet, Portlet, JSF, Spring MVC...). As an advantage of this, the Rendering phase is decentralized and is deported on Client-Side. Consequently, the Server will hold less information in its memory (Stateful UI Tree) and make less computing related to page rendering. As a significant drawback of this new architecture, making more parallel computing (parallel REST calls for each page) makes the multi-threading aspect more important to consider. In fact, a REST call has to be performant and must not hold any potential thread-blocking statement. In addition, the HTTP cache headers has to be considered for almost all REST calls. We have several mime types retrieved through REST calls:

Reviewer: ```suggestion

* `Vue.createApp` to create a new Vue application instead of `new Vue`
* use statement `this.$root.$applicationLoaded()` when all Data is fetched and the UI has been displayed to the End-user.

=== REST API Performances

By increasing the number of REST calls in a single Web-Page, will allow retrieval of information in parallel. This was a major evolution of eXo Platform product architecture introduced in version 6 which will reduce the classic Server-End stateful DOM rendering (that was made using JSP, Servlet, Portlet, JSF, Spring MVC...). As an advantage of this, the Rendering phase is decentralized and is deported on the Client-Side. Consequently, the Server will hold less information in its memory (Stateful UI Tree) and make less computing related to page rendering. As a significant drawback of this new architecture, making more parallel computing (parallel REST calls for each page) makes the multi-threading aspect more important to consider. In fact, a REST call has to be performant and must not hold any potential thread-blocking statement. In addition, the HTTP cache headers have to be considered for almost all REST calls. We have several mime types retrieved through REST calls:
```

* CSS
* Image
* JSON representing a stored data

> For CSS and Images, it's important to consider using a https://developers.google.com/web/fundamentals/performance/get-started/httpcaching-6[long-term cache strategy]. By doing this, the Browser will always retrieve the resource from local cache (Disk or Memory cache) and will not fetch the resource from Server-End again. If the resource can be dynamic, such as illustration on news or Application Center thumbnail, you can add a suffix to the URL that references the `lastModifiedDate` property of the image (https://github.com/Meeds-io/app-center/commit/2b50274f84e9c92c05a71514bf76a802f3f60f99#diff-8a61c4f2ca478f64c19717d95c05298629a81bec5eb62dd4539528970d0a67b1R60[See example])
> 
> For the JSON objects retrieved using REST API, you can consider using a https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag[ETag] to cache objects in browser and to verify that it hasn't been modified since last retrieval. This strategy is very relevant when retrieving information that is not frequently changed, such as an activity that once written, it will not change. By using this strategy, you will avoid to download content from Server-end to save Bandwidth (consequently save time) and even save computing time in Server-End (https://github.com/Meeds-io/social/blob/ffad68f249e4cbfb270e62091fe50c91eb700ecd/component/service/src/main/java/org/exoplatform/social/rest/impl/activity/ActivityRestResourcesV1.java#L345[See example]).

Reviewer: ```suggestion
For CSS and Images, it's important to consider using a https://developers.google.com/web/fundamentals/performance/get-started/httpcaching-6[long-term cache strategy]. By doing this, the Browser will always retrieve the resource from the local cache (Disk or Memory cache) and will not fetch the resource from Server-End again. If the resource can be dynamic, such as illustration on news or Application Center thumbnail, you can add a suffix to the URL that references the `lastModifiedDate` property of the image (https://github.com/Meeds-io/app-center/commit/2b50274f84e9c92c05a71514bf76a802f3f60f99#diff-8a61c4f2ca478f64c19717d95c05298629a81bec5eb62dd4539528970d0a67b1R60[See example])

For the JSON objects retrieved using REST API, you can consider using a https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag[ETag] to cache objects in browser and to verify that it hasn't been modified since the last retrieval. This strategy is very relevant when retrieving information that is not frequently changed, such as an activity that once written, will not change. By using this strategy, you will avoid downloading content from Server-end to save Bandwidth (consequently save time) and even save computing time in Server-End (https://github.com/Meeds-io/social/blob/ffad68f249e4cbfb270e62091fe50c91eb700ecd/component/service/src/main/java/org/exoplatform/social/rest/impl/activity/ActivityRestResourcesV1.java#L345[See example]).
```
